File: 42-trapping-rain-water/42-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        maxLeft = [0] * len(height)
        maxRight = [0] * len(height)
        maxLeft[0] = 0
        maxRight[len(maxRight) - 1] = 0
        for i in range(1, len(height)):
            maxLeft[i] = max(maxLeft[i-1], height[i-1])
        for i in range(len(height) - 2, -1, -1):
            maxRight[i] = max(maxRight[i+1], height[i+1])
        total = 0
        for i in range(0, len(height)):
            water = min(maxLeft[i], maxRight[i]) - height[i]
            if water > 0:
                total += water
        return total

File: 42-trapping-rain-water/42-trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */

var trap = function(height) {
    let maxLeft = height[0];
    let maxRight = height[height.length - 1];
    let l = 0;
    let r = height.length - 1;
    let res = 0;
    
    while (l < r) {
        if (maxLeft < maxRight) {
            l++;
            maxLeft = Math.max(maxLeft, height[l]);
            res += maxLeft - height[l];
        } else {
            r--;
            maxRight = Math.max(maxRight, height[r]);
            res += maxRight - height[r];
        }
    }
    return res;
};